File: Algorithms/46. Permutations/Permute.cs
//Author: Tushar Jaiswal
//Creation Date: 08/26/2018

/*Given a collection of distinct integers, return all possible permutations.
Example:
Input: [1,2,3]
Output:
[
  [1,2,3],
  [1,3,2],
  [2,1,3],
  [2,3,1],
  [3,1,2],
  [3,2,1]
]*/

public class Solution {
    public IList<IList<int>> Permute(int[] nums) {
        List<int> list = new List<int>(nums);
        IList<IList<int>> result = new List<IList<int>>();
        GetPermutations(result, list, new List<int>());
        return result;
    }
    
    private void GetPermutations(IList<IList<int>> result, List<int> nums, List<int> curr)
    {
        if(nums.Count == 0)
        { result.Add(curr); }
        
        foreach(int i in nums)
        {
            List<int> newCurr = new List<int>(curr);
            List<int> newNums = new List<int>(nums);
            newCurr.Add(i);
            newNums.Remove(i);
            GetPermutations(result, newNums, newCurr);
        }
    }
}

File: Algorithms/46. Permutations/Permute.java
//Author: Tushar Jaiswal
//Creation Date: 08/26/2018

/*Given a collection of distinct integers, return all possible permutations.
Example:
Input: [1,2,3]
Output:
[
  [1,2,3],
  [1,3,2],
  [2,1,3],
  [2,3,1],
  [3,1,2],
  [3,2,1]
]*/

class Solution {
    public List<List<Integer>> permute(int[] nums) {
        List<Integer> list = new ArrayList<Integer>();
        for(int i : nums)
        { list.add(i); }
        List<List<Integer>> result = new ArrayList<List<Integer>>();
        getPermutations(result, list, new ArrayList<Integer>());
        return result;
    }
    
    private void getPermutations(List<List<Integer>> result, List<Integer> nums, List<Integer> curr)
    {
        if(nums.size() == 0)
        { result.add(curr); }
        
        for(int i : nums)
        {
            List<Integer> newCurr = new ArrayList<Integer>(curr);
            List<Integer> newNums = new ArrayList<Integer>(nums);
            newCurr.add(i);
            newNums.remove((Integer)i);
            getPermutations(result, newNums, newCurr);
        }
    }
}